### ISSUE TITLE
Does not support unicode strings

### ISSUE THREAD

**Issue by odeits** [cheery/json-algorithm#1]
Tested against https://raw.githubusercontent.com/akrylysov/python-json-benchmark/master/data/twitter.json 

```
Traceback (most recent call last):
  File "json-algorithm/verifier.py", line 227, in <module>
    main()
  File "json-algorithm/verifier.py", line 83, in main
    state = parse_ch(cat, ch, stack, state, ds, ss, es, chart)
  File "json-algorithm/verifier.py", line 114, in parse_ch
    do_action(action, ch, ds, ss, es)
  File "json-algorithm/verifier.py", line 142, in do_action
    val = u"".join(ss)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe5 in position 0: ordinal not in range(128)
```

**Comment by cheery** [comment]
Hi @odeits, That is a gorgeous file!

TL;DR: The fix is to iterate fd.read().decode('utf-8') rather than fd.read()

I replicated your experiment quickly, including this error. To understand what's going on, here's a small explanation of the error.

First of all, put a print ss before the line 142 that crashes, run it again. You should get:

```
['s', 't', 'a', 't', 'u', 's', 'e', 's']
['m', 'e', 't', 'a', 'd', 'a', 't', 'a']
['r', 'e', 's', 'u', 'l', 't', '_', 't', 'y', 'p', 'e']
['r', 'e', 'c', 'e', 'n', 't']
['i', 's', 'o', '_', 'l', 'a', 'n', 'g', 'u', 'a', 'g', 'e', '_', 'c', 'o', 'd', 'e']
['j', 'a']
['c', 'r', 'e', 'a', 't', 'e', 'd', '_', 'a', 't']
['S', 'u', 'n', ' ', 'A', 'u', 'g', ' ', '3', '1', ' ', '0', '0', ':', '2', '9', ':', '1', '5', ' ', '+', '0', '0', '0', '0', ' ', '2', '0', '1', '4']
['i', 'd']
['i', 'd', '_', 's', 't', 'r']
['5', '0', '5', '8', '7', '4', '9', '2', '4', '0', '9', '5', '8', '1', '5', '6', '8', '1']
['t', 'e', 'x', 't']
['@', 'a', 'y', 'm', '0', '5', '6', '6', 'x', ' ', u'\n', u'\n', '\xe5', '\x90', '\x8d', '\xe5', '\x89', '\x8d', ':', '\xe5', '\x89', '\x8d', '\xe7', '\x94', '\xb0', '\xe3', '\x81', '\x82', '\xe3', '\x82', '\x86', '\xe3', '\x81', '\xbf', u'\n', '\xe7', '\xac', '\xac', '\xe4', '\xb8', '\x80', '\xe5', '\x8d', '\xb0', '\xe8', '\xb1', '\xa1', ':', '\xe3', '\x81', '\xaa', '\xe3', '\x82', '\x93', '\xe3', '\x81', '\x8b', '\xe6', '\x80', '\x96', '\xe3', '\x81', '\xa3', '\xef', '\xbc', '\x81', u'\n', '\xe4', '\xbb', '\x8a', '\xe3', '\x81', '\xae', '\xe5', '\x8d', '\xb0', '\xe8', '\xb1', '\xa1', ':', '\xe3', '\x81', '\xa8', '\xe3', '\x82', '\x8a', '\xe3', '\x81', '\x82', '\xe3', '\x81', '\x88', '\xe3', '\x81', '\x9a', '\xe3', '\x82', '\xad', '\xe3', '\x83', '\xa2', '\xe3', '\x81', '\x84', '\xe3', '\x80', '\x82', '\xe5', '\x99', '\x9b', '\xe3', '\x81', '\xbf', '\xe5', '\x90', '\x88', '\xe3', '\x82', '\x8f', '\xe3', '\x81', '\xaa', '\xe3', '\x81', '\x84', u'\n', '\xe5', '\xa5', '\xbd', '\xe3', '\x81', '\x8d', '\xe3', '\x81', '\xaa', '\xe3', '\x81', '\xa8', '\xe3', '\x81', '\x93', '\xe3', '\x82', '\x8d', ':', '\xe3', '\x81', '\xb6', '\xe3', '\x81', '\x99', '\xe3', '\x81', '\xa7', '\xe3', '\x82', '\xad', '\xe3', '\x83', '\xa2', '\xe3', '\x81', '\x84', '\xe3', '\x81', '\xa8', '\xe3', '\x81', '\x93', '\xf0', '\x9f', '\x98', '\x8b', '\xe2', '\x9c', '\xa8', '\xe2', '\x9c', '\xa8', u'\n', '\xe6', '\x80', '\x9d', '\xe3', '\x81', '\x84', '\xe5', '\x87', '\xba', ':', '\xe3', '\x82', '\x93', '\xe3', '\x83', '\xbc', '\xe3', '\x83', '\xbc', '\xe3', '\x83', '\xbc', '\xe3', '\x80', '\x81', '\xe3', '\x81', '\x82', '\xe3', '\x82', '\x8a', '\xe3', '\x81', '\x99', '\xe3', '\x81', '\x8e', '\xf0', '\x9f', '\x98', '\x8a', '\xe2', '\x9d', '\xa4', '\xef', '\xb8', '\x8f', u'\n', 'L', 'I', 'N', 'E', '\xe4', '\xba', '\xa4', '\xe6', '\x8f', '\x9b', '\xe3', '\x81', '\xa7', '\xe3', '\x81', '\x8d', '\xe3', '\x82', '\x8b', '\xef', '\xbc', '\x9f', ':', '\xe3', '\x81', '\x82', '\xe3', '\x81', '\x81', '\xe2', '\x80', '\xa6', '\xe2', '\x80', '\xa6', '\xe3', '\x81', '\x94', '\xe3', '\x82', '\x81', '\xe3', '\x82', '\x93', '\xe2', '\x9c', '\x8b', u'\n', '\xe3', '\x83', '\x88', '\xe3', '\x83', '\x97', '\xe7', '\x94', '\xbb', '\xe3', '\x82', '\x92', '\xe3', '\x81', '\xbf', '\xe3', '\x81', '\xa6', ':', '\xe7', '\x85', '\xa7', '\xe3', '\x82', '\x8c', '\xe3', '\x81', '\xbe', '\xe3', '\x81', '\x99', '\xe3', '\x81', '\x8c', '\xe3', '\x81', '\xaa', '\xf0', '\x9f', '\x98', '\x98', '\xe2', '\x9c', '\xa8', u'\n', '\xe4', '\xb8', '\x80', '\xe8', '\xa8', '\x80', ':', '\xe3', '\x81', '\x8a', '\xe5', '\x89', '\x8d', '\xe3', '\x81', '\xaf', '\xe4', '\xb8', '\x80', '\xe7', '\x94', '\x9f', '\xe3', '\x82', '\x82', '\xe3', '\x82', '\x93', '\xe3', '\x81', '\xae', '\xe3', '\x83', '\x80', '\xe3', '\x83', '\x81', '\xf0', '\x9f', '\x92', '\x96']
```

You can see your 0xe5 byte there!

And you got the same error. Now, ASCII alphabet tends to be 7-bit codes. Python has places where it checks for this. When it hits one you get this kind of UnicodeDecodeError with "Ordinal not in range(128)". The problem is explained in the error message.

```
print (0xe5 not in range(128))
```

Those are utf-8 character points. To get python understand them and treat them as individual characters, you need to decode your utf-8 input file.

**Comment by odeits** [comment]
@cheery Thank you, this resolved the issue. Did you want to add an example of decoding a utf-8 file to the README or provide a test case?

**Comment by cheery** [comment]
@odeits I can give the verifier that functionality.

Going to look on the pretty printing during the weekend. I'll add the file decoding example during that, perhaps.

Btw. Are you going to write a blog post/article on benchmarking this?

**Comment by odeits** [comment]
I don't write a blog. I was just curios and messing around with it. I can
tell you that under pypy it has much better performance but under regular
cPython it is quite slow.
